base/src/dispatcher/dispatcher.ts

```diff
@@ -372,16 +411,97 @@ export class Dispatcher implements ZLUX.Dispatcher {
     
 
 /* what will callback be called with */
-  registerAction(action: Action):void {
-      this.actionsByID.set(action.id,action);
-   }
+  registerAction(action: AbstractAction): void {
+    const { id } = action;
+    const existingAction: AbstractAction | undefined = this.actionsByID.get(id);
+
+    if (existingAction) {
+      this.log.warn(`duplicate actionId ${id}. Replacing`, existingAction, 'with', action);
+    }
+
+    this.actionsByID.set(action.id, action);
+
+    if (action instanceof ActionContainer) {
+      const { children } = action as ActionContainer;
 
+      if (children && children.length > 0) {
+        for (let child of children) {
+          if (child instanceof AbstractAction) {
+            this.registerAction(child);
+          }
+        }
+      }
+    }
+  }
 
-   getAction(recognizer:any):Action|undefined {
+   getAction(recognizer:any):AbstractAction|undefined {
      this.log.debug("actionName "+JSON.stringify(recognizer)+" in "+JSON.stringify(this.actionsByID));
-     return this.actionsByID.get(recognizer.actionID);
+     return this.getActionById(recognizer.actionID);
    }
 
+   getActionById(actionId: string): AbstractAction|undefined {
+     return this.actionsByID.get(actionId);
+   }
+
+  getActions(capabilities: string[] | null, applicationContext: any): ZLUX.ActionLookupResult {
+    const actions: AbstractAction[] = [];
+    const recognizers: RecognitionRule[] = this.getRecognizersForCapabilities(capabilities, applicationContext);
+    const unresolvedActionIds: string[] = [];
+
+    if (recognizers) {
+      for (let recognizer of recognizers) {
+        const action: AbstractAction | undefined = this.getAction(recognizer);
+
+        if (action) {
+          if (action instanceof ActionContainer) {
+            this.resolveActionReferences(action as ActionContainer, unresolvedActionIds);
+          }
+
+          actions.push(action);
+        } else {
+          unresolvedActionIds.push(recognizer.actionID);
+        }
+      }
+    }
+
+    return {
+      actions: actions.length > 0 ? actions : undefined,
+      unresolvedActionIds: unresolvedActionIds.length > 0 ? unresolvedActionIds : undefined
+    };
+  }
+
+  /**
+   * WARNING: recursively modifies the children references
+   * @param actionContainer 
+   */
+  private resolveActionReferences(actionContainer: ActionContainer, unresolvedActionIds: string[]): void {
```

Comment:
Not necessary, but possibility to mark something as not needing to be run twice since this resolution sticks.

Reply:
Given that unresolveActions can change over time (as more actions get registered), I think that getting the caching and returning the right unresolveActionIds might require more thought, so I will skip it for now.

Reply:
Given that unresolveActions can change over time (as more actions get registered), I think that getting the caching and returning the right unresolveActionIds might require more thought, so I will skip it for now.